department-of-veterans-affairs/va.gov-cms
# VACMS-7240: Link validation for protocol-relative links.

## Description

Closes #7240.

## Testing done
Added PHPUnit tests.

## QA steps

As any user with the ability to edit content:
1. Edit a node, e.g. [VAMC System Health Service Whole health at VA Cincinnati health care](https://cms-xc4z49m67roe5vpgv3pxdmlg7avbzekg.ci.cms.va.gov/node/24426/edit)
2. Insert a protocol-relative link into a rich text field.  For instance: `<a href="//www.example.org/">Example Link!</a>`.
3. Attempt to save the document.  You should see:
  - [ ] An error message at the top of the screen:
![Screen Shot 2022-08-29 at 2 07 19 PM](https://user-images.githubusercontent.com/1318579/187269063-c57b92f1-a03f-4e62-8342-d6d27ac6c1bd.png)

  - [ ] An error message at the field:
![Screen Shot 2022-08-29 at 2 07 25 PM](https://user-images.githubusercontent.com/1318579/187269046-deb17c6f-e071-4812-8f2f-e56dc351233d.png)

  - [ ] An error message at the bottom of the field:
![Screen Shot 2022-08-29 at 2 07 29 PM](https://user-images.githubusercontent.com/1318579/187269030-1ff18ace-ded0-42f3-a52a-fbfe43d225ba.png)

4. Edit a media image item, e.g. [August2022.jpg](https://cms-xc4z49m67roe5vpgv3pxdmlg7avbzekg.ci.cms.va.gov/media/image/16769)

5. Insert a protocol-relative link into a plain text field.  For instance, `//www.example.org/`.

6. Attempt to save the media image item.  You should see:
  - [ ] An error message at the top of the screen:
![Screen Shot 2022-08-29 at 2 17 48 PM](https://user-images.githubusercontent.com/1318579/187270641-0e1e94fa-9ee4-4e58-b19b-d564299855c8.png)

  - [ ] An error message at the field:
![Screen Shot 2022-08-29 at 2 17 52 PM](https://user-images.githubusercontent.com/1318579/187270635-9ab234a9-a501-49a0-85c0-551370982f13.png)


  - [ ] An error message at the bottom of the field:
![Screen Shot 2022-08-29 at 2 17 56 PM](https://user-images.githubusercontent.com/1318579/187270630-92839e1f-8267-44cc-9d4f-58b1ae84d7b8.png)

7. Other non-`string_long` plain-text fields can generally use protocol-relative links without issue.  Revision Log message fields, etc should continue to work as before.
